---
title: "16S Amplicon Survey of fecal samples from children with environmental enteropathy, to study their Growth Velocity : Part 2 : Study of Growth Velocity in EED"
author: "Chandni (Chandani) Desai"
output: html_document
editor_options: 
  chunk_output_type: console
---

### Study children with EED - adequate/moderate/poor growth velocity, prior to growth

```{r sample-selection-and-setup}

## SAMPLE SELECTION

## Remove samples with HEALTHY EE status
ps2.healthysamplesremoved <- ps2 %>%
  subset_samples(EE_status_based_on_percL != "No")
ps2.healthysamplesremoved


## SELECT GOOD, MODERATE and POOR growth samples - PRIOR TO GROWTH PERIOD
ps2.goodgrowthpoorgrowth.priorsamplesonly <- ps2.healthysamplesremoved %>%                 
 subset_samples(Growth_Status_Prior %in% c("Adequate", "Poor"))
ps2.goodgrowthpoorgrowth.priorsamplesonly
#otu_table()   OTU Table:         [ 2432 taxa and 44 samples ]

ps2.goodmoderatepoorgrowth.priorsamplesonly <- ps2.healthysamplesremoved %>%                 
 subset_samples(Growth_Status_Prior %in% c("Adequate", "Poor", "Moderate"))
ps2.goodmoderatepoorgrowth.priorsamplesonly
#otu_table()   OTU Table:         [ 2432 taxa and 81 samples ]


ps2.moderategrowthpoorgrowth.priorsamplesonly <- ps2.healthysamplesremoved %>%                 
 subset_samples(Growth_Status_Prior %in% c("Moderate", "Poor"))
ps2.moderategrowthpoorgrowth.priorsamplesonly
#otu_table()   OTU Table:         [ 2432 taxa and 49 samples ]
```

```{r taxon-cleaning}

## Remove Taxa with ZERO reads

ps2.goodgrowthpoorgrowth.priorsamplesonly.bacteria <- prune_taxa(taxa_sums(ps2.goodgrowthpoorgrowth.priorsamplesonly) > 0, ps2.goodgrowthpoorgrowth.priorsamplesonly)
ps2.goodgrowthpoorgrowth.priorsamplesonly.bacteria
#otu_table()   OTU Table:         [ 1135 taxa and 44 samples ]

ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria <- prune_taxa(taxa_sums(ps2.goodmoderatepoorgrowth.priorsamplesonly) > 0, ps2.goodmoderatepoorgrowth.priorsamplesonly)
#otu_table()   OTU Table:         [ 1715 taxa and 81 samples ]


ps2.moderategrowthpoorgrowth.priorsamplesonly.bacteria <- prune_taxa(taxa_sums(ps2.moderategrowthpoorgrowth.priorsamplesonly) > 0, ps2.moderategrowthpoorgrowth.priorsamplesonly)
## 1394 taxa, 49 samples

```


## COMMUNITY COMPOSITION AND RELATIVE ABUNDANCE
```{r community-composition}

## COMMUNITY COMPOSITION AND RELATIVE ABUNDANCE PLOTS
## THREE GROWTH CATEGORIES

threshold=0.01

ps2.goodmoderatepoorgrowth.priorsamplesonly.phylum <- ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria %>%
  tax_glom(taxrank = "Phylum") %>%                         # agglomerate at phylum level
  transform_sample_counts(function(x) {x/sum(x)} ) %>%     # Transform to relative abundance
  psmelt() %>%                                             # Melt to long format
  filter(Abundance > threshold) %>%                        # Filter low abundance taxa
  arrange(Phylum)

p.goodmoderatepoor.communitybarplot.prior <- ggplot(ps2.goodmoderatepoorgrowth.priorsamplesonly.phylum, aes(x = sample_name, y = Abundance, fill = Phylum)) + 
  geom_bar(stat = "identity", width = 0.9, position = "fill") +
  labs(title = "Community Composition by Growth Status", y="Relative Abundance", x="Growth status") +
  facet_wrap(~Growth_Status_Prior, scales = "free_x") +
  theme(legend.position = "bottom") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(axis.text.x = element_blank())
  #theme(axis.title.x = element_blank()) +
p.goodmoderatepoor.communitybarplot.prior

```


## ALPHA DIVERSITY
```{r estimate-richness}

## CALCULATE RICHNESS = THREE GROUPS

## PRIOR
diversity.goodmoderatepoor.prior <- estimate_richness(ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria, measures = c("Observed","Shannon"))
row.names(diversity.goodmoderatepoor.prior) <- c(sub('\\X',"",rownames(diversity.goodmoderatepoor.prior)))
head(diversity.goodmoderatepoor.prior)

sampledata.ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria <- as.data.frame(sample_data(ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria))
ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria.rich <- merge(sampledata.ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria, diversity.goodmoderatepoor.prior, by ="row.names") # merge sample data by row names

write.table(ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria.rich,file="Stunting/ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria.rich",row.names=FALSE,quote = FALSE,sep="\t")

```

```{r plot-alpha-diversity}

# PLOT ALPHA DIVERSITY - PRIOR
ggplot(ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria.rich, aes(age_months,color=Growth_Status_Prior,y=Observed)) + 
  geom_point(size = 3) + 
  geom_line(aes(group = patient_ID)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  #scale_y_continuous(breaks = seq(0, 800, by = 200)) +
  labs(title="Bacterial Richness for each Patient, with Age", y="Observed Richness", x="Age in Months", color="Growth Status")

p.rich.prior <- ggboxplot(ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria.rich, x ="Growth_Status_Prior", y = "Observed", outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(y = "Richness", x="", title="Richness") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  stat_compare_means(label = "p.format", method = "kruskal.test", hide.ns = TRUE)
p.rich.prior

p.sd.prior <- ggboxplot(ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria.rich, x = "Growth_Status_Prior", y = "Shannon", outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(y = "Shannon diversity", x="Height for Age growth status (3months)", title="Shannon Diversity") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  stat_compare_means(label = "p.format", method = "kruskal.test", hide.ns = TRUE)

ggarrange(p.rich.prior, p.sd.prior, nrow = 2, labels = c("A)", "B)"))

```


## BETA DIVERSITY
```{r beta-diversity}

# unifrac
ord.ps2.goodmoderatepoor.uni.prior <- ordinate(ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria, method = "PCoA", distance = "unifrac")

# PCoA plot
p.ord.uni.prior <- plot_ordination(ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria, ord.ps2.goodmoderatepoor.uni.prior, color = "Growth_Status_Prior") +
  geom_point(size=3, alpha = 0.7) +
  scale_color_brewer(palette = "Dark2") +
  ggtitle("UniFrac")
p.ord.uni.prior

# weighted unifrac
ord.ps2.goodmoderatepoor.wuni.prior <- ordinate(ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria, method = "PCoA", distance = "wunifrac")

# PCoA plot
p.ord.wuni.prior <- plot_ordination(ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria, ord.ps2.goodmoderatepoor.wuni.prior, color="Growth_Status_Prior") +
  geom_point(size=3, alpha = 0.7) +
  scale_color_brewer(palette = "Dark2") +
  ggtitle("wUniFrac")
p.ord.wuni.prior

ggarrange(p.ord.uni.prior, p.ord.wuni.prior, nrow = 2, labels = c("A)", "B)"), common.legend = TRUE, legend = "bottom")

```


## BETA DIVERSITY SIGNIFICANCE BY ADONIS
```{r ADONIS}

## HELPFUL LINK
## http://deneflab.github.io/MicrobeMiseq/demos/mothur_2_phyloseq.html#permanova

# Set a random seed so that exact results can be reproduced
set.seed(1000)

# Function to run adonis test on a physeq object and a variable from metadata 
doadonis <- function(physeq, category) {
  bdist <- phyloseq::distance(physeq, "wUnifrac")
  col <- as(sample_data(physeq), "data.frame")[ ,category]
  
  # Adonis test
  adonis.bdist <- adonis(bdist ~ col)
  print("Adonis results:")
  print(adonis.bdist)
  
  # Homogeneity of dispersion test
  betatax = betadisper(bdist,col)
  p = permutest(betatax)
  print("Betadisper results:")
  print(p$tab)
}


doadonis(ps2.goodmoderatepoorgrowth.priorsamplesonly.bacteria, "Growth_Status_Prior")
## ADEQUATE MODERATE POOR : wUnifrac

#          Df SumsOfSqs  MeanSqs F.Model      R2 Pr(>F)
#col        2    0.0808 0.040408 0.76584 0.01926  0.634
#Residuals 78    4.1156 0.052763         0.98074       
#Total     80    4.1964                  1.00000       
#[1] "Betadisper results:"
#          Df     Sum Sq     Mean Sq        F N.Perm Pr(>F)
#Groups     2 0.01281246 0.006406232 1.243553    999  0.283
#Residuals 78 0.40182116 0.005151553       NA     NA     NA

```


## DIFFERENTIAL ABUNDANCE TESTING
```{r prevalence-filter-before-differential-testing}

prevdf.prior <- apply(X = otu_table(ps2.goodgrowthpoorgrowth.priorsamplesonly.bacteria),MARGIN = ifelse(taxa_are_rows(ps2.goodgrowthpoorgrowth.priorsamplesonly.bacteria), yes = 1, no = 2),FUN = function(x){sum(x > 0)})

# Add taxonomy and total read counts to prevdf
prevdf.prior <- data.frame(Prevalence = prevdf.prior, TotalAbundance = taxa_sums(ps2.goodgrowthpoorgrowth.priorsamplesonly.bacteria), tax_table(ps2.goodgrowthpoorgrowth.priorsamplesonly.bacteria))

prevalenceThreshold.prior = 0.03 * nsamples(ps2.goodgrowthpoorgrowth.priorsamplesonly.bacteria)
prevalenceThreshold.prior
#[1] 1.32

# Define which taxa fall within the prevalence threshold
keepTaxa.prior <- rownames(prevdf.prior)[(prevdf.prior$Prevalence >= prevalenceThreshold.prior)]

ntaxa(ps2.goodgrowthpoorgrowth.priorsamplesonly.bacteria)
#[1] 1135
# Apply Prevalence filter
ps2.goodgrowthpoorgrowth.priorsamplesonly.prevfiltered <- prune_taxa(keepTaxa.prior, ps2.goodgrowthpoorgrowth.priorsamplesonly.bacteria)
ntaxa(ps2.goodgrowthpoorgrowth.priorsamplesonly.prevfiltered)
#[1] 570

```

```{r differential-abundance-testing-priortogrowthsamples}

## BACTERIAL TAXA DIFFERENCES BETWEEN ADEQUATE & POOR GROWTH

ds.goodbadpriorsamples <- phyloseq_to_deseq2(ps2.goodgrowthpoorgrowth.priorsamplesonly.prevfiltered, ~bf_at_sampling + Growth_Status_Prior)

levels(ds.goodbadpriorsamples$Growth_Status_Prior)
#[1] "Adequate" "Poor"
## Poor/Adequate

dds.goodbadpriorsamples <- DESeq(ds.goodbadpriorsamples, test="Wald", fitType="local", betaPrior = FALSE)

alpha = 0.05
baseMeanfilter = 10

res.dds.goodbadpriorsamples = results(dds.goodbadpriorsamples, cooksCutoff = FALSE)
# Filter by pAdj
sigtab_dds.goodbadpriorsamples = res.dds.goodbadpriorsamples[which(res.dds.goodbadpriorsamples$padj < alpha & res.dds.goodbadpriorsamples$baseMean > baseMeanfilter), ]

sigtab_dds.goodbadpriorsamples = cbind(as(sigtab_dds.goodbadpriorsamples, "data.frame"), as(ps2.goodgrowthpoorgrowth.priorsamplesonly.prevfiltered@tax_table@.Data[rownames(sigtab_dds.goodbadpriorsamples), ], "matrix"))
head(sigtab_dds.goodbadpriorsamples)
write.table(sigtab_dds.goodbadpriorsamples, file="Stunting/DifferentialAbundance/sigtab_dds.goodgrowth_and_poorgrowth_priorsamples_bf_confounder.txt", sep = "\t", quote = FALSE)


## ###################
## THIS PLOT IS USEFUL
## ###################
x.ds.goodbadpriorsamples = tapply(sigtab_dds.goodbadpriorsamples$log2FoldChange, sigtab_dds.goodbadpriorsamples$Family, function(x) max(x))
x.ds.goodbadpriorsamples = sort(x.ds.goodbadpriorsamples, TRUE)
sigtab_dds.goodbadpriorsamples$Genus = factor(as.character(sigtab_dds.goodbadpriorsamples$Family), levels=names(x.ds.goodbadpriorsamples))
p.deseq.goodbadpriorsamples <- ggplot(sigtab_dds.goodbadpriorsamples, aes(x=Family, y=log2FoldChange, color=Genus)) +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5)) +
  geom_point(size=5, alpha = 0.7) + 
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5)) +
  ggtitle("Taxa Differentially Associated by Poor vs Adequate Growth : Prior to growth period") + 
  geom_hline(yintercept = 0, lwd=1.5)
p.deseq.goodbadpriorsamples

```


## Ground Truth Plots (Normalized Abundance of Individual Differentially Abundant Taxa)
```{r get-normalized-abundance-table}

## MAKE TABLE OF NORMALIZED ABUNDANCES FOR ALL SAMPLES
## Add deseq normalized counts to the phyloseq object (to plot normalized abundance) and convert to a dataframe

replace_counts = function(physeq, dds) {

  dds_counts = counts(dds, normalized = TRUE)
  if (!identical(taxa_names(physeq), rownames(dds_counts))) {
    stop("OTU ids don't match")
  }
  otu_table(physeq) = otu_table(dds_counts, taxa_are_rows = TRUE)
  return(physeq)

}

# Make deseq ready object
ds.all<- phyloseq_to_deseq2(ps2, ~Growth_Status_Prior)


# Calculate geometric means prior to estimate size factors
gm_mean = function(x, na.rm=TRUE){
  exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
}

geoMeans.ds.all <- apply(counts(ds.all), 1, gm_mean)
ds.all <- estimateSizeFactors(ds.all, geoMeans = geoMeans.ds.all)

dds.all <- DESeq(ds.all, fitType = "local", test="Wald")
rlog.all <- replace_counts(ps2, dds.all)

rlog.all <- psmelt(rlog.all)
rlog.all <- rename(rlog.all, c("OTU" = "ASV"))

## This is to recreate the DESeq normalization that takes place before differential abundance testing, to generate normalized counts for taxa abundance

```


```{r plot-taxa-abundance-in-all-growth-categories}

# Get selected taxa and plot

significanttaxa_dataframe_goodbadpriorsamplesbacteria <- as.data.frame(res.dds.goodbadpriorsamples[ which(res.dds.goodbadpriorsamples$padj < alpha & res.dds.goodbadpriorsamples$baseMean > baseMeanfilter), ])

# Check how many differentially abundant taxa came up
nrow(significanttaxa_dataframe_goodbadpriorsamplesbacteria)

significanttaxa_dataframe_goodbadpriorsamplesbacteria <- rownames_to_column(significanttaxa_dataframe_goodbadpriorsamplesbacteria, var = "ASV")

#Annotate with taxonomy
ps2.tax_table <- as.data.frame(tax_table(ps2))
ps2.tax_table <- rownames_to_column(ps2.tax_table, var = "ASV")

significanttaxa_dataframe_goodbadpriorsamplesbacteria.annotated <- (left_join(significanttaxa_dataframe_goodbadpriorsamplesbacteria, ps2.tax_table, by = "ASV"))
colnames(significanttaxa_dataframe_goodbadpriorsamplesbacteria.annotated)

write.table(significanttaxa_dataframe_goodbadpriorsamplesbacteria, file="Stunting/significanttaxa_dataframe_goodbadpriorsamplesbacteria.txt", row.names=FALSE, sep="\t", quote = FALSE)

# Merge significant taxa information with normalized abundance and metadata
abundancedataframe.rlog.pregrowth <- inner_join(significanttaxa_dataframe_goodbadpriorsamplesbacteria, rlog.all, by = "ASV")

nrow(abundancedataframe.rlog.pregrowth)

colnames(abundancedataframe.rlog.pregrowth)

write.table(abundancedataframe.rlog.pregrowth, file="Stunting/significanttaxa_dataframe_goodbadpriorsamplesbacteria_withabundance_andmetadata.txt", row.names=FALSE, sep="\t", quote = FALSE)


## PLOT
## ####
ggplot(subset(abundancedataframe.rlog.pregrowth, Growth_Status_Prior %in% c("Adequate","Moderate", "Poor")), aes(y = Abundance, x = Growth_Status_Prior)) +
  #geom_text(aes(label=patient_ID),hjust=0, vjust=-1) +
  geom_boxplot(width = 0.5,  outlier.shape = NA) +
  geom_jitter(alpha = 0.7, width = 0.15, size = 2) + #aes(color=bf_at_sampling)) +
  scale_y_log10() +
  theme(legend.position = "bottom") +
  facet_wrap(ASV~Genus) +
  labs(title="Normalized Abundance of Bacterial Taxa assoc with Stunting", y="Normalized Abundance(log10)", x="Growth Status", subtitle="Pre growth, at Genus-ASV level")

```