---
  title: "16S Amplicon Survey of fecal samples from children with environmental enteropathy, to study their Growth Velocity : Part 1:Preliminary Analysis"
author: "Chandni (Chandani) Desai"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
  code_folding: hide
#runtime: shiny
editor_options: 
  chunk_output_type: console
---
  
```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=8,
                      fig.height=6,
                      fig.path="./figures/",
                      dev='png',
                      warning=FALSE,
                      message=FALSE)
```


##Load libraries
```{r initiate-environment, include=FALSE}
library("tidyverse")
packageVersion("tidyverse")
library("reshape2")
packageVersion("reshape2")
library("plyr")
packageVersion("plyr")
library("phyloseq")
packageVersion("phyloseq")
library("RColorBrewer")
packageVersion("RColorBrewer")
library("vegan")
packageVersion("vegan")
library("gridExtra")
packageVersion("gridExtra")
library("knitr")
packageVersion("knitr")
library("plotly")
packageVersion("plotly")
library("microbiome")
packageVersion("microbiome")
library("ggpubr")
packageVersion("ggpubr")
library("data.table")
packageVersion("data.table")
library("mgcv")
packageVersion("mgcv")
library("pairwiseAdonis") #https://github.com/pmartinezarbizu/pairwiseAdonis
packageVersion("pairwiseAdonis")
library("DESeq2")
packageVersion("DESeq2")
library("descr")
packageVersion("descr")
library("corrplot")
packageVersion("corrplot")
library("shiny")
packageVersion("shiny")
#setwd("~/Box/Stunting_in_EnvEnteropathy/Scripts_and_DataFiles/")
#setwd("~/Documents/Lori_EE/EnvironmentalEnteropathy/MAIN_EnvEnteropathy/16S/environmental-enteropathy-16S/")
#source("runDifferentialAbundance.R")
```

```{r global-theme-settings, include=FALSE}
# Set global theming
theme_set(theme_bw(base_size = 12,
                   base_family = "Arial"))
```

```{r load-data, include=FALSE}
## Read in data
# Load Phyloseq Object from Preprocessing
ps0 <- readRDS("../data/physeqObjects/ps0.EE_bacterial_holtz.rdp.RDS")
ps0
#phyloseq-class experiment-level object
#otu_table()   OTU Table:         [ 3403 taxa and 157 samples ]
#tax_table()   Taxonomy Table:    [ 3403 taxa by 7 taxonomic ranks ]
#phy_tree()    Phylogenetic Tree: [ 3403 tips and 3401 internal nodes ]
# Load updated mapping data
map <- import_qiime_sample_data("../documents/EE_mapping_16S_v2.txt")
dim(map) # 157 x 45
# Merge ASV table with mapping data
ps0 <- merge_phyloseq(ps0, map)
ps0
# View sample variables & generate basic stats
sample_variables(ps0)
sd(sample_sums(ps0))
#[1] 14519.79
get_taxa_unique(ps0, "Phylum")
#[1] "Bacteroidetes"             "Proteobacteria"            "Firmicutes"                "Actinobacteria"           
# [5] "Spirochaetes"              "Elusimicrobia"             NA                          "Tenericutes"              
# [9] "Fusobacteria"              "Verrucomicrobia"           "Cyanobacteria/Chloroplast" "Euryarchaeota"            
#[13] "Lentisphaerae"             "Chloroflexi"  
ntaxa(ps0)
#[1] 3403
reads.raw <- 
  qplot(rowSums(otu_table(ps0))) +
  xlab("Counts-per-sample") +
  ggtitle("Raw Reads")
reads.raw.log <- qplot(log10(rowSums(otu_table(ps0)))) +
  xlab("Logged counts-per-sample") +
  ggtitle("Log10 Transformed")
ggarrange(reads.raw, reads.raw.log, ncol = 2)
```

## Removal of Duplicate Samples
```{r  duplicate-sample-removal, include=TRUE}
#Remove Duplicate samples
ps0

ps1 <- ps0 %>%
  subset_samples(
    sample_name != "Manary_GCE_Pediatric_Stool_16_LNS_51_initial_NA0000440680" &
    sample_name != "Manary_GCE_Pediatric_Stool_17_LNS_51_midstudy_NA0000494512" &
    sample_name != "Manary_GCE_Pediatric_Stool_18_LNS_51_final_NA0000424860" &
    sample_name != "Manary_GCE_Pediatric_Stool_22_LNS_67_initial_NA0000441943" &
    sample_name != "Manary_GCE_Pediatric_Stool_23_LNS_67_midstudy_NA0000495266" &
    sample_name != "Manary_GCE_Pediatric_Stool_24_LNS_67_final_NA0000421395" &
    sample_name != "Manary_GCE_Pediatric_Stool_55_LNS_203_initial_NA0000425732" &
    sample_name != "Manary_GCE_Pediatric_Stool_56_LNS_203_midstudy_NA0000491667" &
    sample_name != "Manary_GCE_Pediatric_Stool_57_LNS_203_final_NA0000419561" &
    sample_name != "Manary_GCE_Pediatric_Stool_61_LNS_205_initial_NA0000420319" &
    sample_name != "Manary_GCE_Pediatric_Stool_62_LNS_205_midstudy_NA0000441999" &
    sample_name != "Manary_GCE_Pediatric_Stool_63_LNS_205_final_NA0000426287"
   )
ps1 # 3403 x 45

```


```{r factor-adjustments, include=FALSE}
#Factor reordering and renaming
# Examine the way the sample data look now
levels(sample_data(ps1)$stage)
# Reorder to: initial, midstudy, final
sample_data(ps1)$stage <- factor(sample_data(ps1)$stage, levels = c("initial", "midstudy", "final"))
levels(sample_data(ps1)$stage)
sample_data(ps1)$EE_status_based_on_percL
sample_data(ps1)$EE_status_based_on_percL <- factor(sample_data(ps1)$EE_status_based_on_percL, levels = c("No", "Moderate", "Severe"))
sample_data(ps1)$EE_status_based_on_percL
```


```{r set-up-data-and-metadata, include=TRUE}
# Combine sample counts and taxonomy with sample metadata
ss <- sample_sums(ps1)
sd <- as.data.frame(sample_data(ps1))
ss.df <- merge(sd, data.frame("ASV" = ss), by ="row.names") # merge ss with sd by row names. Rename ss to ASVs in the new data frame
```

##Sample Assessment
```{r sample-removal-identification}
# Plot data by main variable (EE status or Growth Status)
#Threshold for the minimum number of acceptable reads.
y = 1000
# x-axis variable you want to examine
x = "age_months"
# label to overlay on the points that are below threshold y. To indicate low abundant samples
label = "Description"
# Plot
p.ss.boxplot<- ggplot(ss.df, aes_string(x, y = "ASV")) + # x is what you assigned it above
  geom_boxplot(outlier.colour="NA", aes(group = stage)) +
  scale_y_log10() +
  geom_hline(yintercept = y, lty = 2) +
  geom_jitter(alpha = 0.6, width = 0.15, size = 2.5) +
  #geom_text(data = subset(ss.df, RSVs <= y), aes_string(x, y="RSVs", label=label), size=2, hjust -1) +# This labels a subset that fall below threshold variable y and labels them with the label variable
  ggtitle("Number of ASV") +
  ylab("ASV (log10)") +
  facet_grid(~EE_status_based_on_percL) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(legend.position = "NULL") +
  theme(axis.title.x = element_blank()) + geom_text(aes(label = sample),hjust = 0, nudge_x = 0.05)
p.ss.boxplot
## NONE OF THE SAMPLES HAD READS BELOW OUR THRESHOLD. Hence, NONE were removed.
```


```{r sampling-timeline}
#Plot sampling timeline
ggplot(sd, aes(age_months,color=EE_status_based_on_percL,y=reorder(factor(patient_ID),age_months))) + 
  geom_point(size = 3) + 
  geom_line(aes(group = patient_ID)) +
  ggtitle("Sampling Timeline") +
  theme(legend.position = "bottom")
```


##Overall sample relationship to evaluate sample outliers
```{r outlier-sample-evaluation}
# Outlier evaluation
out.bray <- ordinate(prune_samples(sample_sums(ps1) >=2, ps1), method = "NMDS", distance = "bray")
p.NMDS.outlier <- plot_ordination(prune_samples(sample_sums(ps1) >=2, ps1), out.bray, color = "EE_status_based_on_percL", axes = c(1,2)) +
  theme_bw() +
  geom_point(size = 2) +
  ggtitle("NMDS of Bray Distances \nOutlier Evaluation")
  #stat_ellipse(type = "norm", geom = "polygon", alpha = 1/10, aes(fill = EE_status_based_on_percL))
p.NMDS.outlier
# NO SAMPLES WERE REMOVED AS OUTLIERS (NO CLEAR OUTLIERS FOUND)
```

##Taxon cleaning 
```{r taxon-cleaning}
# Removing sequences not classified as Bacteria
# Removing sequences classified as either mitochondria or chlorplast
# Remove taxa represented by 0 reads
ps1.remove0readstaxa <- prune_taxa(taxa_sums(ps1) > 0, ps1)
ntaxa(ps1.remove0readstaxa)

ps2 <- ps1.remove0readstaxa %>%
  subset_taxa(
    Kingdom == "Bacteria" &
    Family  != "mitochondria" &
    Class   != "Chloroplast" &
    Phylum != "Cyanobacteria/Chloroplast"
  )
ps2 # 2432 x 145
```

```{r save-ps2-for-R21-follow-up-analyses, eval=FALSE}
# Jan 7, 2020
saveRDS(ps2, "../data/physeqObjects/ps2.RDS")
```

