---
title: "Environmental Enteropathy: Phage Analysis and Figures"
author: "Chandani Desai & Rachel Rodgers"
date: "August 05, 2019"
output: html_document
---

# Phage Analyses for Stunting and Environmental Enteropathy

```{r initiate-environment, include=FALSE}
# Global knitr option
knitr::opts_chunk$set(warning=FALSE, message=FALSE)

# Load libraries
library("phyloseq")
library("RColorBrewer")
library("vegan")
library("gridExtra")
library("knitr")
library("plotly")
library("ggpubr")
library("data.table")
library("microbiome")
library("DESeq2")
library("ggrepel")
library("pheatmap")
library("FSA")
library("gmodels")
library("tidyverse")

source("./shared_R_scripts/Helper_Functions.R")

theme_set(theme_bw(base_size = 14))

# Load Viral Pre-processing Data

load("../data/RDataObjects/EE_Virome_Preprocessing.RData")
physeqVirusFinal # 3035 x 145 (contains all taxa and healthy + EE samples)
physeqSubsetFinal # 1021 x 145 (contains selected phage and healthy + EE samples)
```

## Sample Selection - Growth_Status_Prior, Moderate & Severe EE

Now we will select samples with Adequate, Moderate, or Poor growth based on Growth_Status_Prior.  We will be selecting these samples from physeqSubsetFinal (which contains just the phage taxa of interest).  We are also removing samples that were considered healthy (not having EE) as based on the percent L measurement.

```{r physeqPriorAllEE}
# This also includes "healthy" samples.
levels(sample_data(physeqSubsetFinal)$Growth_Status_Prior)
table(sample_data(physeqSubsetFinal)$Growth_Status_Prior)
# Adequate = 39, Moderate = 42, Poor = 15 (total of 96 samples)

physeqPrior <- physeqSubsetFinal %>% 
  subset_samples(Growth_Status_Prior %in% c("Adequate", "Moderate", "Poor")) %>% 
  RemoveMissingTaxa()
physeqPrior # 1008 x 96

table(sample_data(physeqPrior)$EE_status_based_on_percL)
# No = 13, Moderate = 39, Severe = 44 (83 samples have EE)

physeqPriorAllEEPhage <- physeqPrior %>% 
  subset_samples(EE_status_based_on_percL != "No") %>% 
  RemoveMissingTaxa()
physeqPriorAllEEPhage # 1004 x 83

table(sample_data(physeqPriorAllEEPhage)$Growth_Status_Prior)
# Adequate = 33, Moderate = 37, Poor = 13

phageSamples <- data.frame("phage_samples" = sample_names(physeqPriorAllEEPhage))
phageSamplesUnique <- phageSamples %>% 
  mutate("unique_id" = str_extract(phage_samples, 
                                   pattern = "-[:digit:]+-LNS-?[:digit:]+-"),
         "unique_id" = str_remove_all(unique_id, pattern = "-"))

#saveRDS(physeqPriorAllEEPhage, "../data/physeqObjects/physeqPriorAllEEPhage.RDS")
```

```{r euk-phage-sample-compare, eval=FALSE}
eukFamSamples <- readRDS("./data/RDataObjects/eukFamSamples.RDS") # 91
phageSamplesVec <- phageSamplesUnique$unique_id # 87

eukFamSamples[!eukFamSamples %in% phageSamplesVec]

phageSamplesVec[!phageSamplesVec %in% eukFamSamples]
```

## Phage Community Composition by Growth Status Prior

```{r phage-community-composition}
phageAbundanceDF <- MakeAbundanceDF(physeqPriorAllEEPhage,
                                    taxRank = "Rank1", abundanceFilter = 0.01)
phageAbundanceDF$Rank1 <- as.character(phageAbundanceDF$Rank1)

phageAbundanceDF <- phageAbundanceDF %>% 
  mutate("Taxon" = ifelse(Rank0 == "other_phages",
                          yes = "other phages", no = Rank1))

phageLevels <- unique(phageAbundanceDF$Taxon)
phageLevels <- phageLevels[phageLevels != "other phages"]
phageLevelsArranged <- phageLevels[order(phageLevels)]
phageLevelsArranged <- c(phageLevelsArranged, "other phages")

phageAbundanceDF$Taxon <- factor(phageAbundanceDF$Taxon,
                                 levels = phageLevelsArranged)

phageCommComp <- ggplot(phageAbundanceDF,
                        aes(x = sample_name, y = Abundance, fill = Taxon)) +
  geom_bar(stat = "identity", width = 1, position = "fill") +
  labs(y = "Relative Abundance") +
  facet_wrap(~ Growth_Status_Prior, scales = "free_x") +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        #axis.title.y = element_text(size = 12),
        strip.text = element_text(size = 10),
        legend.position = "bottom",
        legend.title = element_blank(),
        legend.text = element_text(size = 9),
        legend.box.margin = margin(-10, -10, -10, -10),
        legend.key.size = unit(0.4, "cm"),
        legend.spacing.x = unit(0.05, "cm"),
        panel.grid = element_blank()) +
  scale_fill_brewer(palette = "Set1")
phageCommComp
```

## Phage Richness Regression by Age in Months (EE Samples Only)

```{r CalculateAlphaDiversity}
CalculateAlphaDiversity <- function(physeqObj) {
  alphaDivDF <- estimate_richness(physeqObj,
                                  measures = c("Observed", "Shannon"))
  # Change periods in alphaDivDF back to hyphens so we can merge w/sample data
  row.names(alphaDivDF) <- c(gsub("\\.", "-", rownames(alphaDivDF)))
  
  sampleDataDF <- as(sample_data(physeqObj), "data.frame")
  
  alphaDivDF <- merge(alphaDivDF, sampleDataDF, by = "row.names")
  return(alphaDivDF)
}
```

```{r alphaDivDFList}
phageAlphaDiv <- CalculateAlphaDiversity(physeqPriorAllEEPhage)

#write.table(phageAlphaDiv, "../analyses/phageRichnessAndDiversity_191209.txt",
            #row.names = FALSE, quote = FALSE, sep = "\t")

```

```{r phagel-richness-regression}
phageRichRegression <- lm(Observed ~ age_months, data = phageAlphaDiv)
summary(phageRichRegression)
```

```{r phageRichRegressionPlot}
phageRichRegressionPlot <- ggplot(phageRichRegression,
                                  aes(x = age_months, y = Observed)) + 
  geom_point(size = 1.5) + 
  stat_smooth(method = lm) +
  labs(y="Phage Richness", x="Age in months") +
  annotate("text", x = 20, y = 600, label = "paste(italic(R) ^ 2, \" = 0.027\")", 
           parse = TRUE) +
  theme(axis.title = element_text(size = 10),
        panel.grid = element_blank())
set.seed(2137091)
phageRichRegressionPlot
```

Wilcox p-values for alpha div by growth status:

```{r wilcox-on-alpha-by-growth-status}
growthCategories <- c("Adequate" = "Adequate", "Moderate" = "Moderate",
                      "Poor" = "Poor")

phageAlphaDiv$Growth_Status_Prior <- as.character(phageAlphaDiv$Growth_Status_Prior)

RunWilcox <- function(df, growthCat, alphaCat) {
  
  initialVec <- df %>% 
    dplyr::filter(Growth_Status_Prior == growthCat, stage == "initial") %>% 
    dplyr::pull(!!sym(alphaCat))
  
  midstudyVec <- df %>% 
    dplyr::filter(Growth_Status_Prior == growthCat, stage == "midstudy") %>% 
    dplyr::pull(!!sym(alphaCat))
    
  res <- wilcox.test(initialVec, midstudyVec)
  return(res)
}

wilcoxResultsRich <- map(.x = growthCategories,
                         .f = RunWilcox,
                         df = phageAlphaDiv,
                         alphaCat = "Observed")
wilcoxResultsRich

wilcoxResultsDiv <- map(.x = growthCategories,
                        .f = RunWilcox,
                        df = phageAlphaDiv,
                        alphaCat = "Shannon")
wilcoxResultsDiv
```

## Phage Alpha and Beta Diversity Plots

```{r labels-for-panels-C-D}
# for panels C, D
timepointLabels <- c("Growth Period 1", "Growth Period 2")
names(timepointLabels) <- c("initial", "midstudy")
```

```{r unpair-timepoint1-timepoint2}
# timepoint 1: stage = "initial"
phageAlphaDivT1 <- filter(phageAlphaDiv, stage == "initial")

# timepoint 2: stage = "midstudy"
phageAlphaDivT2 <- filter(phageAlphaDiv, stage == "midstudy")
```

```{r PlotAlphaDiv}
PlotAlphaDiv <- function(df, index, yLabel) {
  set.seed(2908795)
  ggplot(data = df, aes_string(x = "Growth_Status_Prior", y = index)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(alpha = 0.7, width = 0.15, size = 1.5) +
    ylab(yLabel) +
    stat_compare_means(label = "p.format", method = "kruskal.test",
                       hide.ns = TRUE, label.x.npc = 0.5,
                       size = 3.5) +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_text(angle = 45, hjust = 1))
  }
```

```{r phageRichUnpaired}
phageRichUnpaired <- PlotAlphaDiv(df = phageAlphaDiv,
                                  index = "Observed",
                                  yLabel = "Richness") +
  facet_wrap(~ stage, labeller = labeller(stage = timepointLabels)) +
  theme(panel.grid = element_blank())
  
set.seed(21233406)
phageRichUnpaired
```

Dunn's post-hoc test for unpaired richness:

```{r richness-post-hoc}
dunnsRichnessT1 <- dunnTest(phageAlphaDivT1$Observed, phageAlphaDivT1$Growth_Status_Prior)
dunnsRichnessT1$res

dunnsRichnessT2 <- dunnTest(phageAlphaDivT2$Observed, phageAlphaDivT2$Growth_Status_Prior)
dunnsRichnessT2$res
```

```{r phageDivUnpaired}
phageDivUnpaired <- PlotAlphaDiv(df = phageAlphaDiv,
                                 index = "Shannon",
                                 yLabel = "Shannon Diversity") +
  facet_wrap(~ stage, labeller = labeller(stage = timepointLabels)) +
  ylim(0, 5.5) +
  theme(panel.grid = element_blank())
set.seed(21233406)
phageDivUnpaired
```

```{r div-post-hoc}
dunnsDivT1 <- dunnTest(phageAlphaDivT1$Shannon, phageAlphaDivT1$Growth_Status_Prior)
dunnsDivT1$res

dunnsDivT2 <- dunnTest(phageAlphaDivT2$Shannon, phageAlphaDivT2$Growth_Status_Prior)
dunnsDivT2$res
```

### Phage MDS of Bray Distances

```{r PlotOrdination}
PlotOrdination <- function(df, eigenVec, plotTitle = NULL) {
  ggplot(df, aes_string(x = "Axis.1", y = "Axis.2", color = "Growth_Status_Prior")) +
    xlab(paste0("PCoA 1 (", round(100*eigenVec[1], digits = 1), "%)")) +
    ylab(paste0("PCoA 2 (", round(100*eigenVec[2], digits = 1), "%)")) +
    geom_point(size = 2, alpha = 0.7) +
    scale_color_manual(values = c("red", "blue", "green")) +
    ggtitle(plotTitle) +
    coord_fixed(sqrt(eigenVec[2]/eigenVec[1])) +
    theme(legend.position = "right",
          legend.title = element_blank(),
          legend.text = element_text(size = 10),
          axis.title = element_text(size = 10),
          plot.title = element_text(hjust = 0.5, size = 10),
          panel.grid = element_blank())
}
```

```{r mdsBrayUnpaired}
physeqPriorAllEEPhageT1 <- subset_samples(physeqPriorAllEEPhage,
                                          stage == "initial") # 1004 x 44
physeqPriorAllEEPhageT2 <- subset_samples(physeqPriorAllEEPhage,
                                          stage == "midstudy") # 1004 x 39

brayList <- list("T1" = physeqPriorAllEEPhageT1, "T2" = physeqPriorAllEEPhageT2)

set.seed(40072100)
adonisValues <- map(.x = brayList,
                    .f = RunAdonis,
                    distance = "bray", category = "Growth_Status_Prior")

set.seed(40072100)
ordBrayValues <- map(.x = brayList,
                     .f = phyloseq::ordinate,
                     distance = "bray", method = "MDS")

eigenValuesT1 <- ordBrayValues$T1$values$Relative_eig
eigenValuesT2 <- ordBrayValues$T2$values$Relative_eig

ordinationDFs <- pmap(.l = list(physeq = brayList,
                                ordination = ordBrayValues),
                      .f = plot_ordination,
                      justDF = TRUE)

set.seed(40072100)
ordinationPlotsUnpaired <- pmap(.l = list(df = ordinationDFs,
                                          eigenVec = list("T1" = eigenValuesT1,
                                                          "T2" = eigenValuesT2),
                                          plotTitle = list("T1" = "Growth Period 1",
                                                           "T2" = "Growth Period 2")),
                                .f = PlotOrdination)

ordinationPlotsUnpaired
```

### Arrang Phage Panel

```{r, fig.width=8.5, fig.height=11}
fig5TopRow <- ggarrange(phageCommComp, phageRichRegressionPlot,
                        labels = c("A", "B"), ncol = 2)

fig5MiddleRow <- ggarrange(phageRichUnpaired,
                           phageDivUnpaired,
                           labels = c("C", "D"), ncol = 2)

fig5BottomRow <- ggarrange(ordinationPlotsUnpaired$T1 +
                             annotate("text", x = -0.3, y = 0.3, 
                                      label = "p = 0.063", size = 3.5),
                           ordinationPlotsUnpaired$T2 +
                             annotate("text", x = -0.2, y = -0.3,
                                      label = "p = 0.365", size = 3.5),
                           labels = "E", ncol = 2,
                           common.legend = TRUE,
                           legend = "bottom")


phagePanel <- ggarrange(fig5TopRow, fig5MiddleRow, fig5BottomRow, nrow = 3,
                        heights = c(0.75, 0.75, 1))
phagePanel
```

Phage panel scaling

```{r phage-panel-scaling, eval=FALSE}
tiff("../figures/second_rebuttal_figures/Figure5.tiff", 
     units = "in", width = 8.5, height = 11, res = 300)
phagePanel
dev.off()

pdf("../figures/second_rebuttal_figures/Figure5.pdf", width = 8.5, height = 11)
phagePanel
dev.off()
```

### DESeq on Phage

First need to prevalence filter - remove taxa present in only one sample.

```{r phage-prevalence}
#----- prevalence filtering (taxa  present in > 1 sample) -----#
phagePrevVec <- apply(X = otu_table(physeqPriorAllEEPhage),
                      MARGIN = ifelse(taxa_are_rows(physeqPriorAllEEPhage),
                                      yes = 1, no =2),
                      FUN = function(x) {sum(x > 0)})

phageAbdVec <- taxa_sums(physeqPriorAllEEPhage)

phageTaxTable <- as(tax_table(physeqPriorAllEEPhage), "matrix")

phagePrevDF <- data.frame(Prevalence = phagePrevVec,
                          TotalAbundance = phageAbdVec,
                          phageTaxTable)

keepPhageDESeq <- phagePrevDF %>% 
  rownames_to_column(var = "taxon") %>% 
  filter(Prevalence > 1) %>% 
  pull(taxon) # 934 keepers

physeqPhageFilt <- prune_taxa(keepPhageDESeq, physeqPriorAllEEPhage)
physeqPhageFilt # 934 x 83

any(sample_sums(physeqPhageFilt) == 0) # FALSE

#----- Remove Moderate Samples -----#
#physeqPhageNoModerate <- physeqPhageFilt %>% 
  #subset_samples(Growth_Status_Prior != "Moderate") %>% 
  #RemoveMissingTaxa()
#physeqPhageNoModerate # 915 x 46

#----- How Many Samples in Each Growth Category? -----#
#sampleDataPhageDESeq <- as(sample_data(physeqPhageNoModerate), "data.frame")
#sampleDataPhageDESeq %>% 
  #group_by(Growth_Status_Prior) %>% 
  #summarize(count = n())

#----- DESeq2 Analysis -----#
# replace any missing values within bf_at_sampling with "unknown"
sample_data(physeqPhageFilt)$bf_at_sampling <- as.character(sample_data(physeqPhageFilt)$bf_at_sampling)
sample_data(physeqPhageFilt)$bf_at_sampling[is.na(sample_data(physeqPhageFilt)$bf_at_sampling)] <- "unknown"
# make class of patient_ID factor instead of numeric
sample_data(physeqPhageFilt)$patient_ID <- as.factor(sample_data(physeqPhageFilt)$patient_ID)

dds <- phyloseq_to_deseq2(physeqPhageFilt,
                          design = ~ bf_at_sampling + patient_ID + Growth_Status_Prior)

ddsAnalysis <- DESeq(dds)

#----- DESeq2 Results -----#
ddsResAP <- results(ddsAnalysis, 
                    contrast = c("Growth_Status_Prior", "Adequate", "Poor"))
ddsResAM <- results(ddsAnalysis,
                    contrast = c("Growth_Status_Prior", "Adequate", "Moderate"))
ddsResMP <- results(ddsAnalysis,
                    contrast = c("Growth_Status_Prior", "Moderate", "Poor"))

deseqResultsList <- list("AP" = ddsResAP, "AM" = ddsResAM, "MP" = ddsResMP)

# For each data set, generate results table using 0.1 as the significance threshold
deseqResTableList <- pmap(.l = list(ddsResults = deseqResultsList),
                          .f = GenerateDESeqResultsTable,
                          physeq = physeqPhageFilt,
                          sigThreshold = 0.1)

#----- Extract, Merge Significant Taxa -----#
# From the deseqResTableList, extract significant results
deseqSigTables <- map(.x = deseqResTableList,
                     .f = ~ dplyr::filter(.x, Significant == TRUE))

deseqSigOTUs <- map(.x = deseqSigTables, .f = ~ dplyr::select(.x, OTU))

# Get all unique significant OTUs 
allSigOTUs <- Reduce(f = function(df1, df2) {merge(x = df1, y = df2,
                                                   by = "OTU", all = TRUE)},
                     x = deseqSigOTUs)

# Add taxonomy to allSigOTUs
taxTable <- data.frame(as(tax_table(physeqPhageFilt), "matrix")) %>% 
  rownames_to_column(var = "OTU")

allSigOTUs <- merge(allSigOTUs, taxTable, by = "OTU", all = FALSE)
```

```{r annotations-for-heatmap, eval=FALSE}
# row annotations (rank 1 and 3)
allSigOTUs <- allSigOTUs %>% 
  dplyr::mutate("Taxon" = paste(Rank1, Rank3))

rowAnno <- allSigOTUs %>% 
  dplyr::select(OTU, Taxon) %>% 
  column_to_rownames("OTU")
```

```{r extract-normalized-abundance-values}
# replace the counts in the physeq object so they're normalized like the dds object
replace_counts = function(physeq, dds) {
  dds_counts = DESeq2::counts(dds, normalized = TRUE)
  if (!identical(taxa_names(physeq), rownames(dds_counts))) {
    stop("OTU ids don't match")
  }
  otu_table(physeq) = otu_table(dds_counts, taxa_are_rows = TRUE)
  return(physeq)
}

physeqPhageFiltNormalized <- replace_counts(physeq = physeqPhageFilt,
                                            dds = ddsAnalysis)

phageNormalizedMelt <- psmelt(physeqPhageFiltNormalized)

# filter for just the significant taxa
phageNormalizedMeltSigOTUs <- phageNormalizedMelt %>% 
  filter(OTU %in% allSigOTUs$OTU)

# add pseudocount
pseudoNorm <- phageNormalizedMeltSigOTUs %>% 
  mutate(pseudo_norm = 1 + Abundance)

# average normalized OTU counts across growth categories
avgSigOTUNormCounts <- pseudoNorm %>%
  group_by(Growth_Status_Prior, OTU) %>% 
  summarize(avg_norm_count = log2(mean(pseudo_norm)))

# pivot wider to get correct matrix form for heatmap (rows as taxa, columns as 
#   growth categories)
heatmapNormDF <- avgSigOTUNormCounts %>% 
  pivot_wider(names_from = Growth_Status_Prior,
              values_from = avg_norm_count) %>% 
  column_to_rownames(var = "OTU")

heatmapNormMtx <- as.matrix(heatmapNormDF)

# Get row labels
otuLUT <- rowAnno %>% 
  rownames_to_column() %>% 
  deframe()

heatmapRowNames <- data.frame("OTU" = row.names(heatmapNormMtx)) %>% 
  mutate("Taxon" = otuLUT[OTU])

phageHeatmap <- pheatmap(heatmapNormMtx, show_colnames = TRUE,
                         show_rownames = TRUE,
                         annotation_row = rowAnno,
                         cluster_cols = FALSE,
                         cluster_rows = TRUE,
                         labels_row = heatmapRowNames$Taxon,
                         cellwidth = 100,
                         cellheight = 100)
```

```{r save-deseq-results-tables, eval=FALSE}
renamedSigTables <- deseqSigTables
names(renamedSigTables) <- c("phage_Adequate-Poor",
                             "phage_Adequate-Moderate",
                             "phage_Moderate-Poor")
write_xlsx(renamedSigTables,
           path = "../data/phage_DESeq_Tables.xlsx")
```

```{r save-heatmap, eval=FALSE}
pdf("../figures/second_rebuttal_figures/Figure6_raw.pdf",
    width = 11, height = 6)
phageHeatmap
dev.off()
```

# Removed (Related to Old DESeq Analyses)

```{r phage-deseq-results-plot, fig.width=7.5, fig.height=6, eval=FALSE}
phageSigASVDF <- deseqPhageResultsTable %>% 
  filter(Significant == TRUE)

# how many significant taxa were identified?
length(unique(phageSigASVDF$OTU)) # 13

phageSigASVDFModified <- phageSigASVDF %>% 
  mutate("TaxaName" = paste(Rank1, Rank2, Rank3, Rank4),
         "stderror_lower" = log2FoldChange - lfcSE,
         "stderror_upper" = log2FoldChange + lfcSE)

# Fix taxa names by generating a long-to-short name LUT
sigPhageNamesLong <- phageSigASVDFModified$TaxaName
sigPhageNamesShort <- c("unclassified Myoviridae Bacillus phage BCD7",
                        "unclassified phage",
                        "unclassified Podoviridae Pseudomonas phage AF",
                        "unclassified Lactococcus phage",
                        "unclassified Siphoviridae Croceibacter phage P2559Y",
                        "unclassified Myoviridae Rhodothermus phage RM378",
                        "unclassified Clostridium phage phiCT19406C",
                        "unclassified Siphoviridae Clostridium phage phiCD111",
                        "unclassified Siphoviridae",
                        "unclassified Siphoviridae Listeria phage P40",
                        "unclassified Siphoviridae Enterococcus phage phiFL4A",
                        "unclassified Myoviridae Psychrobacter phage pOW20-A",
                        "unclassified Microviridae phi-CA82")

sigPhageNamesLUT <- sigPhageNamesShort
names(sigPhageNamesLUT) <- sigPhageNamesLong

phageNameTable <- enframe(sigPhageNamesLUT)
#write.table(phageNameTable, "../analyses/phageNameTable.txt",
            #row.names = FALSE, sep = "\t", quote = FALSE)

# also save a table of OTUs and the short phage names for correlation plotting
phageRenamingTable <- phageSigASVDFModified %>% 
  dplyr::select(OTU, TaxaName) %>% 
  mutate("short_phage_name" = sigPhageNamesLUT[TaxaName])

write.table(phageRenamingTable, "../analyses/phageRenamingTable.txt",
            row.names = FALSE, sep = "\t", quote = FALSE)


phageSigASVDFModified <- phageSigASVDFModified %>% 
  mutate("corrected_TaxaName" = sigPhageNamesLUT[TaxaName]) %>% 
  arrange(desc(log2FoldChange))

correctedTaxaNameLevels <- phageSigASVDFModified$corrected_TaxaName
phageSigASVDFModified$corrected_TaxaName <- factor(phageSigASVDFModified$corrected_TaxaName,
                                                   levels = correctedTaxaNameLevels)

phageDESeqResultsPlotFinal <- ggplot(phageSigASVDFModified,
                                     aes(x = log2FoldChange, 
                                         y = corrected_TaxaName)) +
  geom_point(size=2) +
  geom_vline(xintercept = 0, alpha = 0.5, linetype = 2) +
  geom_errorbarh(aes(xmin = stderror_lower, xmax = stderror_upper, height=0.1)) +
  labs(x = "log2 Fold Change", y="",subtitle = "Adequate <------> Poor") +
  theme_bw() +
  theme(plot.subtitle = element_text(hjust = 0.5))
phageDESeqResultsPlotFinal
```

```{r phage-deseq-tiff, eval=FALSE}
# figure 6
pdf("../figures/Figure6.pdf",
    height = 8.5, width = 11)
phageDESeqResultsPlotFinal
dev.off()

tiff("../figures/Figure6.tiff", 
     units = "in", width = 7.5, height = 6, res = 300)
phageDESeqResultsPlotFinal
dev.off()
```

Ground truth plots of all differential phage taxa.

```{r fig.width=11, fig.height=8.5, eval=FALSE}
phageSigASVVec <- phageSigASVDF$OTU

sigPhageNormalizedAbdDF <- psViralNormalizedMelt %>% 
  filter(ASV %in% phageSigASVVec)

otuLUT <- as.character(phageSigASVDFModified$corrected_TaxaName)
names(otuLUT) <- phageSigASVDFModified$OTU

sigPhageNormalizedAbdDFModified <- sigPhageNormalizedAbdDF %>% 
  filter(Growth_Status_Prior %in% c("Adequate", "Moderate", "Poor")) %>% 
  mutate(corrected_TaxaName = otuLUT[ASV])

any(is.na(sigPhageNormalizedAbdDFModified$corrected_TaxaName)) # FALSE

# apply levels to this normalized abundance df so order of facets match
#   order of taxa in the deseq results plot
sigPhageNormalizedAbdDFModified$corrected_TaxaName <- factor(sigPhageNormalizedAbdDFModified$corrected_TaxaName,
                                                             levels = rev(correctedTaxaNameLevels))

unique(sigPhageNormalizedAbdDFModified$corrected_TaxaName)

# remove Moderate Growth_Status_Prior
sigPhageNormalizedAbdDFModified <- sigPhageNormalizedAbdDFModified %>% 
  filter(Growth_Status_Prior != "Moderate") %>% 
  droplevels()

# update levels for breastfeeding
levels(sigPhageNormalizedAbdDFModified$bf_at_sampling)
sigPhageNormalizedAbdDFModified$bf_at_sampling <- factor(sigPhageNormalizedAbdDFModified$bf_at_sampling,
                                                         labels = c("Not Breastfed",
                                                                    "Breastfed"))
```

```{r groundTruthPlotsFacet, fig.width = 11, fig.height = 7.5, eval=FALSE}
groundTruthPlotsFacet <- ggplot(sigPhageNormalizedAbdDFModified,
                                aes_string(x = "Growth_Status_Prior",
                                           y = "Abundance")) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  geom_jitter(alpha = 0.7, width = 0.15, size = 1.5) +
  scale_y_log10() +
  facet_wrap(~ corrected_TaxaName,
             labeller = labeller(corrected_TaxaName = label_wrap_gen(10)),
             ncol = 6) +
  theme_bw() +
  theme(legend.position = "none",
        axis.text.x = element_text(angle = 45, hjust = 1, size = 10),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size = 10),
        axis.title = element_text(size = 10),
        strip.text = element_text(size = 8))
groundTruthPlotsFacet
```

```{r select-sipho-for-assembly, eval=FALSE}
microviridaeAbdDF <- sigPhageNormalizedAbdDFModified %>% 
  filter(corrected_TaxaName == "unclassified Microviridae phi-CA82") %>% 
  arrange(desc(Abundance))

# does any sample have all the phage in it?
collapsedAbdDF <- sigPhageNormalizedAbdDFModified %>% 
  select(Sample, Abundance, corrected_TaxaName) %>%
  filter(Abundance > 0) %>% 
  unique() %>% 
  mutate_if(is.factor, as.character)

test <- collapsedAbdDF %>% group_by(Sample) %>% summarise(n_distinct(corrected_TaxaName))
```

```{r save-phage-ground-truth, eval=FALSE}
tiff("../figures/FigureS1.tiff", 
     units = "in", width = 11, height = 7.5, res = 300)
groundTruthPlotsFacet
dev.off()

pdf("../figures/FigureS1.pdf",
    width = 11, height = 7.5)
groundTruthPlotsFacet
dev.off()
```

For each differentially abundant phage, we want to know what percent of samples in the not breastfed versus breastfed category are positive for the presence of that taxon, within growth category.

```{r percent-samples-positive-for-taxa, eval=FALSE}
sigPhageAbdTable <- sigPhageNormalizedAbdDFModified %>% 
  dplyr::select(corrected_TaxaName, Sample, Growth_Status_Prior, bf_at_sampling, Abundance) %>% 
  mutate("Present" = ifelse(Abundance > 0, yes = TRUE, no = FALSE))

# how many total breastfed vs non-breastfed samples are there?
samplesByBreastfeeding <- sigPhageAbdTable %>% 
  dplyr::select(Sample, bf_at_sampling) %>% 
  unique()
table(samplesByBreastfeeding$bf_at_sampling) # 40 not breastfed, 14 bf

# double check this looks correct
sigPhageAbdTable %>% 
  group_by(corrected_TaxaName, bf_at_sampling) %>% 
  summarize("total_samples" = n())

sigPhageSummaryTable <- sigPhageAbdTable %>% 
  group_by(corrected_TaxaName, bf_at_sampling, Growth_Status_Prior) %>% 
  summarize("total_samples" = n(), "total_true" = sum(Present)) %>% 
  mutate("percent_positive_for_taxon" = (total_true/total_samples) * 100)

sigPhageSummaryTableFinal <- arrange(sigPhageSummaryTable,
                                    corrected_TaxaName, Growth_Status_Prior,
                                    bf_at_sampling)
```

```{r save-summary-table, eval=FALSE}
write.table(sigPhageSummaryTableFinal, "../analyses/sigPhageSummaryTableFinal.txt",
            quote = FALSE, sep = "\t", row.names = FALSE)
```

## Extract Normalized Abundances for Significant Phage from DESeq Results

For each significant phage detected by DESeq, extract the abundance values from psViralNormalizedMelt.

```{r significantPhageAbundanceDF, eval=FALSE}
significantPhageASVs <- pull(phageSigASVDFModified, "OTU")
length(significantPhageASVs)
significantPhageAbundanceDF <- filter(psViralNormalizedMelt,
                                      ASV %in% significantPhageASVs)

saveRDS(significantPhageAbundanceDF, "../data/RDataObjects/significantPhageAbundanceDF.RDS")
```

```{r save-image}
save.image("../data/RDataObjects/EE_Phage_Analysis_and_Figures.RData")
```

### Session Info

```{r session-info}
writeLines(capture.output(sessionInfo()), 
           "EE_Virome_Phage_Analysis_and_Figures_session_info.txt")
Sys.Date()
getwd()
sessionInfo()
```

